src/TestExplorer/LSPTestDiscovery.ts

```diff
@@ -29,104 +34,118 @@ import { LanguageClientManager } from "../sourcekit-lsp/LanguageClientManager";
-                                location: client.protocol2CodeConverter.asLocation(func.location),
-                            };
-                        });
-                    return {
-                        name: item.name,
-                        location: client.protocol2CodeConverter.asLocation(item.location),
-                        functions: functions,
-                    };
-                });
-            return classes;
+            const workspaceTestCaps = client.initializeResult?.capabilities.experimental;
+            if (!workspaceTestCaps) {
+                throw new Error("textDocument/tests requests not supported");
+            }
+            const workspaceTestsCaps = workspaceTestCaps[workspaceTestsRequest.method];
+
+            // Only use the lsp for this request if it supports the
+            // workspace/tests method, and is at least version 2.
+            if (workspaceTestsCaps?.version >= 2) {
+                const tests = await client.sendRequest(workspaceTestsRequest, {}, token);
+                const testsInWorkspace = tests.filter(item =>
+                    isPathInsidePath(
+                        client.protocol2CodeConverter.asLocation(item.location).uri.fsPath,
+                        workspaceRoot.fsPath
+                    )
+                );
+
+                return this.transform(client, swiftPackage, testsInWorkspace);
+            } else {
+                throw new Error("workspace/tests requests not supported");
+            }
         });
     }
+
+    /**
+     * Convert from `LSPTestItem[]` to `TestDiscovery.TestClass[]`,
+     * updating the format of the location.
+     */
+    private transform(
+        client: LanguageClient,
+        swiftPackage: SwiftPackage,
+        input: LSPTestItem[]
+    ): TestDiscovery.TestClass[] {
+        return input.map(item => {
+            const location = client.protocol2CodeConverter.asLocation(item.location);
+            const id = this.transformId(item, location, swiftPackage);
+            return {
+                ...item,
+                id: id,
+                location: location,
+                children: this.transform(client, swiftPackage, item.children),
+            };
+        });
+    }
+
+    /**
+     * If the test is an XCTest, transform the ID provided by the LSP from a
+     * swift-testing style ID to one that XCTest can use. This allows the ID to
+     * be used to tell to the test runner (xctest or swift-testing) which tests to run.
+     */
+    private transformId(
+        item: LSPTestItem,
+        location: vscode.Location,
+        swiftPackage: SwiftPackage
+    ): string {
+        // XCTest: Target.TestClass/testCase
+        // swift-testing: TestClass/testCase()
+        //                TestClassOrStruct/NestedTestSuite/testCase()
+
+        let id: string = item.id;
+        if (item.style === "XCTest") {
+            const target = swiftPackage
+                .getTargets(TargetType.test)
```

Comment:
suggestion: maybe we fetch this list in "transform" so we're not filtering each time the map callback is called

Reply:
Any given test might be in a different test target, based on its location. This means we need to check each test's target. A cached lookup of locations to test targets may improve performance a bit but there wont likely be a lot of test targets to look through in a given project.